mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp

```diff
@@ -0,0 +1,99 @@
+//===- RockTuningImpl.cpp - tuning API implementation ----*-===//
+//
+// Part of the rocMLIR Project, under the Apache License v2.0 with LLVM
+// Exceptions. See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// Copyright (c) 2022 Advanced Micro Devices INc.
+//===----------------------------------------------------------------------===//
+//
+// This file implements the tuning interfaces
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Rock/Tuning/RockTuning.h"
+
+namespace mlir {
+namespace rock {
+
+// Brute-force search in incremental order
+void createGemmTuningRangeBF(struct TunableParams *newSpace,
+                             RockGemmWrapperInterface gemmOp) {
+
+  OpBuilder b(gemmOp.getContext());
+  GemmFeatures currentFeatures = gemmOp.getGemmFeatures();
+  if (bitEnumContainsAll(currentFeatures, GemmFeatures::mfma)) {
+    // XDLOPS
+    for (uint32_t gemmMPerBlock : ValidRangeXdlopsGemmParams[0]) {
+      for (uint32_t gemmNPerBlock : ValidRangeXdlopsGemmParams[1]) {
+        for (uint32_t gemmKPerBlock : ValidRangeXdlopsGemmParams[2]) {
+          for (uint32_t gemmMPerWave : ValidRangeXdlopsGemmParams[3]) {
+            for (uint32_t gemmNPerWave : ValidRangeXdlopsGemmParams[4]) {
+              for (uint32_t gemmKPack : ValidRangeXdlopsGemmParams[5]) {
+                XdlopsGemmParamsAttr gemmParams =
+                    b.getAttr<XdlopsGemmParamsAttr>(
+                        gemmKPerBlock, gemmMPerBlock, gemmNPerBlock, gemmKPack,
+                        gemmMPerWave, gemmNPerWave);
+                newSpace->tuningRange.push_back(
+                    gemmParams.cast<RockTuningParamAttrInterface>());
```

Comment:
This seems extremely inefficient. Down sides:
 - Time complexity is very bad. The initialization time forced to be amount of tuning space.
 - Space complexity is very bad. We brute force duplicated a lot of information.

Reply:
I'm not trying to have something optimal this time because we don't know which way is the best at the moment.

Firstly, I'd like to store the parameters in the storage so that we can change the order at any time. This will help exploring quick tuning in the future and we can get something else once we figure out what to do.
With that, there is no faster way to get full tuning space, This is exactly the same we're currently doing in MIOpenDriver code.

Also, this is only a part of the internal implementation, we can change this in the future without changing the API itself. For example, we can embed the tuning parameter generator as a function in the tuning space struct under the same API.

Reply:
You can *probably* ignore this after I have understood your intention in `Quick` tuning space which may be sorted and not linear, that's why we may have to expand this whole thing.